Not able to remove expired lease for Access Token

Hi, 

I have seen many stale leases for GCP access token, and I am not able to clean them up:
```
[ERROR] secrets.system.system_6aeff124: lease revocation failed: lease_id=gcp/token/tsk8s-dde/3614e5d1-6cc2-c7cb-75eb-bd86127facd7 error="failed to revoke entry: resp: &logical.Response{Secret:<nil>, Auth:<nil>, Data:map[string]interface {}{"error":"googleapi: got HTTP response code 400 with body: {\n  \"error\": \"invalid_token\"\n}"}, Redirect:"", Warnings:[]string(nil), WrapInfo:(*wrapping.ResponseWrapInfo)(nil)} err: <nil>"
```

By the way, should these leases be removed automatically since they are not refreshable?

Thanks

--- a/plugin/secrets_access_token.go
+++ b/plugin/secrets_access_token.go
@@ -16,6 +16,8 @@ import (
 	"golang.org/x/oauth2/google"
 	"google.golang.org/api/googleapi"
 	"google.golang.org/api/iam/v1"
+	"strings"
+	"log"
 )
 
 const (
@@ -80,6 +82,11 @@ func (b *backend) secretAccessTokenRenew(ctx context.Context, req *logical.Reque
 	return logical.ErrorResponse("short-term access tokens cannot be renewed - request new access token instead"), nil
 }
 
+func isInvalidTokenErr(err error) bool {
+	gErr, ok := err.(*googleapi.Error)
+	return ok && gErr.Code == 400 && strings.Contains(gErr.Body, "invalid_token")
+}
+
 func secretAccessTokenRevoke(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
 	tokenRaw, ok := req.Secret.InternalData["access_token"]
 	if !ok {
@@ -87,11 +94,17 @@ func secretAccessTokenRevoke(ctx context.Context, req *logical.Request, d *frame
 	}
 
 	resp, err := http.Get(revokeAccessTokenEndpoint + fmt.Sprintf("?token=%s", url.QueryEscape(tokenRaw.(string))))
-	if err != nil {
-		return logical.ErrorResponse(fmt.Sprintf("revoke returned error: %v", err)), nil
+	if err == nil {
+		err = googleapi.CheckResponse(resp)
 	}
-	if err := googleapi.CheckResponse(resp); err != nil {
-		return logical.ErrorResponse(err.Error()), nil
+
+	if err != nil {
+		// Token may have already expired on server; ignore if OAuth server returns error.
+		if req.Secret.ExpirationTime().Before(time.Now()) && isInvalidTokenErr(err) {
+			log.Printf("[WARNING] Manual token revocation failed - this is usually because token has already been invalidated. Ignored error: %v", err)
+		} else {
+			return logical.ErrorResponse(err.Error()), nil
+		}
 	}
 
 	return &logical.Response{

--- a/plugin/secrets_access_token.go
+++ b/plugin/secrets_access_token.go
@@ -16,7 +16,6 @@ import (
 	"golang.org/x/oauth2/google"
 	"google.golang.org/api/googleapi"
 	"google.golang.org/api/iam/v1"
-	"log"
 	"strings"
 )
 
@@ -36,7 +35,7 @@ func secretAccessToken(b *backend) *framework.Secret {
 			},
 		},
 		Renew:  b.secretAccessTokenRenew,
-		Revoke: secretAccessTokenRevoke,
+		Revoke: b.secretAccessTokenRevoke,
 	}
 }
 
@@ -83,11 +82,13 @@ func (b *backend) secretAccessTokenRenew(ctx context.Context, req *logical.Reque
 }
 
 func isInvalidTokenErr(err error) bool {
-	gErr, ok := err.(*googleapi.Error)
-	return ok && gErr.Code == 400 && strings.Contains(gErr.Body, "invalid_token")
+	if gerr, ok := err.(*googleapi.Error); ok {
+		return gerr.Code == 400 && strings.Contains(gerr.Body, "invalid_token")
+	}
+	return false
 }
 
-func secretAccessTokenRevoke(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+func (b *backend) secretAccessTokenRevoke(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
 	tokenRaw, ok := req.Secret.InternalData["access_token"]
 	if !ok {
 		return nil, fmt.Errorf("secret is missing token internal data")
@@ -101,10 +102,15 @@ func secretAccessTokenRevoke(ctx context.Context, req *logical.Request, d *frame
 	if err != nil {
 		// Token may have already expired on server; ignore if OAuth server returns error.
 		if req.Secret.ExpirationTime().Before(time.Now()) && isInvalidTokenErr(err) {
-			log.Printf("[WARNING] Manual token revocation failed - this is usually because token has already been invalidated. Ignored error: %v", err)
-		} else {
-			return logical.ErrorResponse(err.Error()), nil
+			invalidTokenWarn := fmt.Sprintf("manual token revocation failed because token has already been invalidated. ignoring error: '%v'", err)
+			b.Logger().Warn(invalidTokenWarn)
+
+			return &logical.Response{
+				Warnings: []string{invalidTokenWarn},
+			}, nil
 		}
+
+		return logical.ErrorResponse(err.Error()), nil
 	}
 
 	return &logical.Response{

--- a/plugin/secrets_access_token.go
+++ b/plugin/secrets_access_token.go
@@ -16,8 +16,8 @@ import (
 	"golang.org/x/oauth2/google"
 	"google.golang.org/api/googleapi"
 	"google.golang.org/api/iam/v1"
-	"strings"
 	"log"
+	"strings"
 )
 
 const (

